sklearn/preprocessing/_discretization.py

```diff
@@ -259,8 +268,14 @@ def transform(self, X):
             Data in the binned space.
         """
         check_is_fitted(self)
+        # if output dtype is not defined we make it consistent with input dtype
+        if self.dtype is not None:
+            Xt = check_array(X, copy=True, dtype=self.dtype)
+        else:
+            Xt = check_array(X, copy=True, dtype=FLOAT_DTYPES[:2])
+        if 'onehot' in self.encode:
+            self._encoder.dtype = Xt.dtype
```

Comment:
One thing is that we cannot change self in transform. If we have to do it temporary, we need to revert it before the end of this function, say at the very end,

```py
   dtype_init = self._encoder.dtype 
   self._encoder.dtype = Xt.dtype
   Xt_enc = self._encoder.transform(Xt)
   # revert the initial dtype to avoid modifying self.
   self._encoder.dtype = dtype_init
   return Xt_enc
```

It's a bit ugly but I don't really see a workaround.

Reply:
Done.

Reply:
@rth does it makes sense to do 

```python 
    dtype_init = self._encoder.dtype 
    self._encoder.dtype = Xt.dtype
    try:
        Xt_enc = self._encoder.transform(Xt)
    finally:
        # revert the initial dtype to avoid modifying self.
        self._encoder.dtype = dtype_init
    return Xt_enc
```
It would ensure that we modify the dtype even if transform fail but I am not sure this is actually useful?